pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -324,47 +363,67 @@ func (r *Reconciler) lookupLatestClusterPolicySnapshot(ctx context.Context, crp
 		klog.ErrorS(err, "Invalid clusterPolicySnapshots", "clusterResourcePlacement", crpKObj)
-		return nil, -1, controller.NewUnexpectedBehaviorError(err)
+		return nil, -1, nil, controller.NewUnexpectedBehaviorError(err)
 	}
 	// When there are no active snapshots, find the one who has the largest policy index.
-	if err := r.Client.List(ctx, snapshotList, client.MatchingLabels{fleetv1beta1.CRPTrackingLabel: crp.Name}); err != nil {
-		klog.ErrorS(err, "Failed to list all clusterPolicySnapshots", "clusterResourcePlacement", crpKObj)
-		return nil, -1, controller.NewAPIServerError(false, err)
+	sortedList, err := r.listSortedClusterSchedulingPolicySnapshots(ctx, crp)
+	if err != nil {
+		return nil, -1, nil, err
 	}
-	if len(snapshotList.Items) == 0 {
+
+	if len(sortedList.Items) == 0 {
 		// The policy index of the first snapshot will start from 0.
-		return nil, -1, nil
+		return nil, -1, sortedList, nil
 	}
-	index := -1           // the index of the cluster policy snapshot array
-	lastPolicyIndex := -1 // the assigned policy index of the cluster policy snapshot
-	for i := range snapshotList.Items {
-		policyIndex, err := parsePolicyIndexFromLabel(&snapshotList.Items[i])
+	latestSnapshot := &sortedList.Items[len(sortedList.Items)-1]
+	policyIndex, err := parsePolicyIndexFromLabel(latestSnapshot)
+	if err != nil {
+		klog.ErrorS(err, "Failed to parse the policy index label", "clusterResourcePlacement", crpKObj, "clusterPolicySnapshot", klog.KObj(latestSnapshot))
+		return nil, -1, nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	return latestSnapshot, policyIndex, sortedList, nil
+}
+
+// listSortedClusterSchedulingPolicySnapshots returns the policy snapshots sorted by the policy index.
+func (r *Reconciler) listSortedClusterSchedulingPolicySnapshots(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) (*fleetv1beta1.ClusterSchedulingPolicySnapshotList, error) {
+	snapshotList := &fleetv1beta1.ClusterSchedulingPolicySnapshotList{}
+	crpKObj := klog.KObj(crp)
+	if err := r.Client.List(ctx, snapshotList, client.MatchingLabels{fleetv1beta1.CRPTrackingLabel: crp.Name}); err != nil {
+		klog.ErrorS(err, "Failed to list all clusterSchedulingPolicySnapshots", "clusterResourcePlacement", crpKObj)
+		// CRP controller needs a scheduling policy snapshot watcher to enqueue the CRP request.
+		// So the snapshots should be read from cache.
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	sort.Slice(snapshotList.Items, func(i, j int) bool {
+		ii, err := parsePolicyIndexFromLabel(&snapshotList.Items[i])
 		if err != nil {
-			klog.ErrorS(err, "Failed to parse the policy index label", "clusterPolicySnapshot", klog.KObj(&snapshotList.Items[i]))
-			return nil, -1, controller.NewUnexpectedBehaviorError(err)
+			klog.ErrorS(controller.NewUnexpectedBehaviorError(err), "Failed to parse the policy index label", "clusterSchedulingPolicySnapshot", klog.KObj(&snapshotList.Items[i]))
```

Comment:
If an error occurs here, the sort result is unclear. I am not sure if simply logging the error and leave the default value for comparison is the best behavior here. It's better to have a check beforehand.

Reply:
done, returned the unexpected error instead of error swallowing.